tfhe/src/integer/server_key/radix_parallel/scalar_comparison.rs

```diff
@@ -0,0 +1,785 @@
+        let message_max = message_modulus - 1;
+
+        // The idea is that we will sum chunks of blocks until carries are full
+        // then we compare the sum with 0.
+        //
+        // If all blocks were 0, the sum will be zero
+        // If at least one bock was not zero, the sum won't be zero
+        let num_additions_to_fill_carry = (total_modulus - message_max) / message_max;
+        let is_equal_to_zero = self.key.generate_accumulator(|x| {
+            if matches!(comparison_type, ZeroComparisonType::Equality) {
+                u64::from((x % total_modulus as u64) == 0)
+            } else {
+                u64::from((x % total_modulus as u64) != 0)
+            }
+        });
+
+        lhs.par_chunks(num_additions_to_fill_carry)
+            .map(|chunk| {
+                let mut sum = chunk[0].clone();
+                for other_block in &chunk[1..] {
+                    self.key.unchecked_add_assign(&mut sum, other_block);
+                }
+                self.key
+                    .apply_lookup_table_assign(&mut sum, &is_equal_to_zero);
+                sum
+            })
+            .collect::<Vec<_>>()
+    }
+
+    // Given a slice of scalar values, and a total_modulus
+    // where  each scalar value is < total_modulus
+    //
+    // This will return a vector of size `total_modulus`,
+    // where for each index, the vec contains either
+    // - `None` if fhe scalar was not present in the slice,
+    // - or `Some` lookuptable that allows to compare a shortint block to the scalar value at this
+    //  index
+    //
+    //
+    //  E.g.
+    //  - input slice: [0, 2],
+    //  - total_modulus: 4,
+    //  returns -> [Some(LUT(|x| x == 0)), None, Some(LUT(|x| x == 2), None]
+    fn create_scalar_comparison_luts<F>(
+        &self,
+        scalar_blocks: &[u8],
+        total_modulus: usize,
+        comparison_fn: F,
+    ) -> Vec<Option<LookupTableOwned>>
+    where
+        F: Fn(u8, u8) -> bool + Sync,
+    {
+        // One lut per scalar block
+        // And only generate a lut for scalar block
+        // actually present
+
+        // let mut scalar_comp_luts = vec![None; total_modulus];
+        // for scalar_block in scalar_blocks.iter().copied() {
+        //     if scalar_comp_luts[scalar_block as usize].is_some() {
+        //         // The LUT for this scalar has already been generated
+        //         continue;
+        //     }
+        //     let lut = self
+        //         .key
+        //         .generate_accumulator(|x| u64::from(x == scalar_block as u64));
+        //     scalar_comp_luts[scalar_block as usize] = Some(lut);
+        // }
+        // scalar_comp_luts
+
+        let scalar_comp_luts = (0..total_modulus)
+            .map(|_| RwLock::new(None))
+            .collect::<Vec<_>>();
+
+        scalar_blocks.par_iter().copied().for_each(|scalar_block| {
+            let lock = &scalar_comp_luts[scalar_block as usize];
```

Comment:
There is unique in itertolos but not the stdlib, maybe I can try using it and see how it performs.

The reduce the locking a bit, maybe we could use `try_read`, and if we get `WouldBlockError` that means it the lock is 
write locked and so a thread is already generating the lut, no need to go further.

But I also need to re-check that the parallel lut gen brings perf improvements compared to the non parallel version
(I think it did but very very minor)

Reply:
not sure it's worth it, just wondered if it could have been interesting, or maybe just allocate the LUTs once in a once cell somewhere ? not sure it's desirable